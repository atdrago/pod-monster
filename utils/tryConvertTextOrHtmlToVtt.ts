import rehypeParse from 'rehype-parse';
import rehypeStripHtml from 'rehype-strip-html';
import { unified } from 'unified';

import { secondsToFormattedTime } from './date';

function timestampToFormattedTime(timestamp: string): string {
  // Remove brackets
  let cueTiming = timestamp.substring(1, timestamp.length - 1);

  // Add milliseconds if necessary
  if (!/\d{2}:\d{2}:\d{2}.\d{3}/.test(cueTiming)) {
    cueTiming = `${cueTiming}.000`;
  }

  return cueTiming;
}

export async function tryConvertTextOrHtmlToVtt(
  textOrHtml: string,
  podcastDurationSeconds?: number | null
): Promise<string | undefined> {
  // Strip HTML
  const rawText = String(
    await unified().use(rehypeParse).use(rehypeStripHtml).process(textOrHtml)
  );

  // Get start/end indexes of bracket-style timestamps (ex. [00:00:00] or [00:00:00.000])
  const timestampRegExp = /\[\d{2}:\d{2}:\d{2}(.\d{3})?\]/gm;
  let match;
  const timestampMatches = [];

  while ((match = timestampRegExp.exec(rawText)) !== null) {
    timestampMatches.push({
      end: timestampRegExp.lastIndex,
      start: match.index,
    });
  }

  if (timestampMatches.length === 0) {
    return undefined;
  }

  // Build up list of cues
  const cues = [];

  for (let i = 0; i < timestampMatches.length; i++) {
    const timestampMatch = timestampMatches[i];
    const nextTimestampMatch = timestampMatches[i + 1];

    const timestamp = rawText.substring(
      timestampMatch.start,
      timestampMatch.end
    );
    const text = rawText
      .substring(timestampMatch.end, nextTimestampMatch?.start)
      .trim();

    const nextTimestamp = nextTimestampMatch
      ? rawText.substring(nextTimestampMatch.start, nextTimestampMatch.end)
      : undefined;

    let end = null;

    if (nextTimestamp) {
      end = timestampToFormattedTime(nextTimestamp);
    } else if (podcastDurationSeconds) {
      end = secondsToFormattedTime(podcastDurationSeconds, {
        includeMilliseconds: true,
      });
    }

    // Exclude the cue if `text` is empty, or if `end` could not be calculated
    if (text && end) {
      cues.push({
        end,
        start: timestampToFormattedTime(timestamp),
        text,
      });
    }
  }

  // Make VTT file
  const vtt = [
    'WEBVTT',
    ...cues.map((c) => `${c.start} --> ${c.end}\n${c.text}`),
    '', // Important we end with \n\n
  ].join('\n\n');

  return vtt;
}
