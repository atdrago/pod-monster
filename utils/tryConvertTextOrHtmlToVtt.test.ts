import { tryConvertTextOrHtmlToVtt } from './tryConvertTextOrHtmlToVtt';

it('should return undefined when not parsable', async () => {
  const vtt = await tryConvertTextOrHtmlToVtt('foo bar', 60);
  expect(vtt).toEqual(undefined);
});

it('should convert text to vtt', async () => {
  // Snippet from - https://share.transistor.fm/s/0ba4b425/transcript.txt
  const textInput = `[00:00:00] **Cassidy Williams:** Previously on Remotely Interesting...

[00:00:03] **Jason Lengstorf:** For everyone who's not watching this live Tara just opened her mouth at the end of that joke.

[00:00:10] **Cassidy Williams:** Hello and welcome to Remotely Interesting.

[00:00:12] **Phil Hawksworth:** This is remotely interesting.

[00:00:14] **Divya Tagtachian:** Well, that seems a little presumptuous.

[00:00:16] **Sarah Drasner:** No no no, that's the name of the show.

[00:00:26] **Cassidy Williams:** Hello? Do you value things? I doubt it because I don't. Just kidding. We probably value things. And the company you work for, probably values things to, uh, does that align? Oh, no, but we're joined by the only guest that has ever come back to our show. Thuy Doan as well as some other nerds. Why don't you all introduce yourselves?

[00:00:55] **Cassidy Williams:** Fine, Jason. We also have Phil and Charlie who will [00:01:00] not introduce themselves either. And so we're going to get a wonderful little sample size of opinions on this matter and learn from each other's experiences. So hello everyone.

[00:01:14] **Cassidy Williams:** So does your company have values? That's that's my, that's my straight up question because I feel like a lot of companies talk about it, but I, I have seen my fair share of companies who talk the talk, but don't necessarily walk the walk, um, toilets.

[00:01:31] **Jason Lengstorf:** Well, you're going to talk about it and you're going to be about it

[00:01:34] **Cassidy Williams:** Word, huh?

[00:01:36] **Thuy Doan:** Oh snap. So I actually, I'm pretty sure, sure. I know my company's values, but I had to Google it just in case, cause I didn't want to say the wrong thing. Okay. So my company, they have four values, uh, being user obsessed, radical candor, growth mindset, and extreme ownership. I feel like three of them [00:02:00] are pretty straightforward, but in case people don't know what radical candor is.`;

  const vtt = await tryConvertTextOrHtmlToVtt(textInput, 305);
  expect(vtt).toMatchSnapshot();
});

it('should convert html to vtt', async () => {
  // Snippet from - https://feeds.buzzsprout.com/1538779/10139859/transcript
  const textInput = `<body>
    <h1><!--block-->Podland 24/02</h1>
    <p><!--block-->[00:00:00]</p>
    <p><!--block-->[00:00:00] Welcome to Podland the last word in podcasting use. It's the 24th of February, 2022. I'm James. Cridlin the editor of pod news.net.</p>
    <p><!--block-->[00:00:08] And I'm Sam Sethi, the MD of river radio, the podcast, first radio station going live on dab on the 1st of March.</p>
  </body>`;

  const vtt = await tryConvertTextOrHtmlToVtt(textInput, 7025);

  expect(vtt).toMatchSnapshot();
});

it('should exclude last transcript cue if duration is not specified', async () => {
  // Snippet from - https://feeds.buzzsprout.com/1538779/10139859/transcript
  const textInput = `<body>
    <h1><!--block-->Podland 24/02</h1>
    <p><!--block-->[00:00:00]</p>
    <p><!--block-->[00:00:00] Welcome to Podland the last word in podcasting use. It's the 24th of February, 2022. I'm James. Cridlin the editor of pod news.net.</p>
    <p><!--block-->[00:00:08] And I'm Sam Sethi, the MD of river radio, the podcast, first radio station going live on dab on the 1st of March.</p>
  </body>`;

  const vtt = await tryConvertTextOrHtmlToVtt(textInput);

  expect(vtt).toMatchSnapshot();
});
