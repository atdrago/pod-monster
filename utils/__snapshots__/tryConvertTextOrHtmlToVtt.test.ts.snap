// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should convert html to vtt 1`] = `
"WEBVTT

00:00:00.000 --> 00:00:08.000
Welcome to Podland the last word in podcasting use. It's the 24th of February, 2022. I'm James. Cridlin the editor of pod news.net.

00:00:08.000 --> 01:57:05.000
And I'm Sam Sethi, the MD of river radio, the podcast, first radio station going live on dab on the 1st of March.

"
`;

exports[`should convert text to vtt 1`] = `
"WEBVTT

00:00:00.000 --> 00:00:03.000
**Cassidy Williams:** Previously on Remotely Interesting...

00:00:03.000 --> 00:00:10.000
**Jason Lengstorf:** For everyone who's not watching this live Tara just opened her mouth at the end of that joke.

00:00:10.000 --> 00:00:12.000
**Cassidy Williams:** Hello and welcome to Remotely Interesting.

00:00:12.000 --> 00:00:14.000
**Phil Hawksworth:** This is remotely interesting.

00:00:14.000 --> 00:00:16.000
**Divya Tagtachian:** Well, that seems a little presumptuous.

00:00:16.000 --> 00:00:26.000
**Sarah Drasner:** No no no, that's the name of the show.

00:00:26.000 --> 00:00:55.000
**Cassidy Williams:** Hello? Do you value things? I doubt it because I don't. Just kidding. We probably value things. And the company you work for, probably values things to, uh, does that align? Oh, no, but we're joined by the only guest that has ever come back to our show. Thuy Doan as well as some other nerds. Why don't you all introduce yourselves?

00:00:55.000 --> 00:01:00.000
**Cassidy Williams:** Fine, Jason. We also have Phil and Charlie who will

00:01:00.000 --> 00:01:14.000
not introduce themselves either. And so we're going to get a wonderful little sample size of opinions on this matter and learn from each other's experiences. So hello everyone.

00:01:14.000 --> 00:01:31.000
**Cassidy Williams:** So does your company have values? That's that's my, that's my straight up question because I feel like a lot of companies talk about it, but I, I have seen my fair share of companies who talk the talk, but don't necessarily walk the walk, um, toilets.

00:01:31.000 --> 00:01:34.000
**Jason Lengstorf:** Well, you're going to talk about it and you're going to be about it

00:01:34.000 --> 00:01:36.000
**Cassidy Williams:** Word, huh?

00:01:36.000 --> 00:02:00.000
**Thuy Doan:** Oh snap. So I actually, I'm pretty sure, sure. I know my company's values, but I had to Google it just in case, cause I didn't want to say the wrong thing. Okay. So my company, they have four values, uh, being user obsessed, radical candor, growth mindset, and extreme ownership. I feel like three of them

00:02:00.000 --> 00:05:05.000
are pretty straightforward, but in case people don't know what radical candor is.

"
`;

exports[`should exclude last transcript cue if duration is not specified 1`] = `
"WEBVTT

00:00:00.000 --> 00:00:08.000
Welcome to Podland the last word in podcasting use. It's the 24th of February, 2022. I'm James. Cridlin the editor of pod news.net.

"
`;
